text/0000-must-not-suspend-lint.md

```diff
@@ -0,0 +1,186 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Provide a lint that can be attached to structs to let the compiler know that this struct can not be held across an await boundary.
+
+```rust
+#[must_not_suspend = "Your error message here"]
+struct MyStruct {}
+```
+
+This struct if held across an await boundary would cause a warn-by-default warning:
+
+```rust
+async fn foo() {
+  let my_struct = MyStruct {};
+  my_async_op.await;
+  println!("{:?}", my_struct);
+}
+```
+
+The compiler might output something along the lines of:
+
+```
+warning: `MyStruct` should not be held across an await point.
+```
+
+Example use cases for this lint:
+
+- `MutexGuard` holding this across a yield boundary in a single threaded executor could cause deadlocks. In a multi-threaded runtime the resulting future would become `!Send` which will stop the user from spawning this future and causing issues. But in a single threaded runtime which accepts `!Send` futures deadlocks could happen.
+
+- The same applies to other such synchronization primitives such as locks from `parking-lot`.
+
+- `tracing::Span` has the ability to enter the span via the `tracing::span::Entered` guard. While entering a span is totally normal, during an async fn the span only needs to be entered once before the `.await` call, which might potentially yield the execution.
+
+- Any RAII guard might possibly create unintended behavior if held across an await boundary.
+
+This lint will enable the compiler to warn the user that the code could produce unforeseen side effects. Some examples of this are:
+
+- [`std::sync::MutexGuard`](https://doc.rust-lang.org/std/sync/struct.MutexGuard.html)
+- [`tracing::span::Entered`](https://docs.rs/tracing/0.1.15/tracing/span/struct.Entered.html)
+
+This will be a best effort lint to signal the user about unintended side-effects of using certain types across an await boundary.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `must_not_suspend` attribute is used to issue a diagnostic warning when a value is not "used". It can be applied to user-defined composite types (structs, enums and unions), functions and traits.
+
+The `must_not_suspend` attribute may include a message by using the [`MetaNameValueStr`] syntax such as `#[must_not_suspend = "example message"]`.  The message will be given alongside the warning.
+
+When used on a user-defined composite type, if a value exists across an await point, then this lint is violated.
+
+
+```rust
+#[must_not_suspend = "Your error message here"]
+struct MyStruct {}
+
+async fn foo() {
+  let my_struct = MyStruct {};
+  my_async_op.await;
+  println!("{:?}", my_struct);
+}
+```
+
+When used on a function, if the value returned by a function is held across an await point, this lint is violated.
```

Comment:
What should happen if the value returned implements `Copy`?

What should happen in the following two examples?

```rust
#[must_not_suspend]
fn foo() -> i32 { 5i32 }

async fn foo() {
  let mut baz = 0;
  {
    let bar = foo();
    baz = bar;
    drop(bar);
  }
  my_async_op.await;
  println!("{:?}", baz);
  // Ok or not?
}
```

```rust
#[must_not_suspend]
fn foo() -> i32 { 5i32 }

fn qux(a: i32) -> i32 { a }

async fn foo() {
  let bar = qux(foo());
  my_async_op.await;
  println!("{:?}", bar);
  // Ok or not?
}
```

Reply:
Your first example doesn't compile.

Reply:
The relevant question is whether the value is used again after an `await`. `Copy` doesn't affect that at all.

As for your examples, I don't know of a use case for supporting the attribute on functions and would prefer to only support it on types for now.

Reply:
> attribute on functions

I was only using the functions to obtain a value with `must_not_suspend`.

> `Copy` doesn't affect that at all

The thing about `Copy` is that when you have one value, you can easily end up with two, three, or more values resulting from copying the original value.

My question is basically will the copies will inherit the `must_not_suspend`?

- If they do, are we just going to make all functions that take `must_not_suspend` values return results that are also `must_not_suspend`? Otherwise, would we make `Copy` special and why?
- if they don't, then it will become easy to shoot oneself in the foot again if one doesn't realize that they copied a value before an `await`

The same logic could apply to `Clone`, but it's not much of a problem there because having to write out the `.clone()` makes things obvious.

> my answer for `Copy` is the same as for the other examples

So I guess that means the `must_not_suspend` DOES inherit across unmarked functions from the parameters to the return value?

Reply:
> My question is basically will the copies will inherit the must_not_suspend?

If we limit the behavior to just the types, then the copies will inherit the behavior, since any instance of a type marked `#[must_not_suspend]` should not be suspended across await points. I think of it as less inheriting behavior, and more applying behavior applicable to an instance of the type. I'll update the doc to reflect this.